build/bazel/remote/execution/v2/remote_execution.proto

```diff
@@ -1065,14 +1065,15 @@ message OutputFile {
 // [Directory][build.bazel.remote.execution.v2.Directory] protos in a
 // single directory Merkle tree, compressed into one message.
 message Tree {
-  // The root directory in the tree.
-  Directory root = 1;
+  reserved 1, 2; // Used for removed fields in an earlier version of the API.
+
+  // The digest of the root directory in the tree.
+  Digest rootDigest = 3;
 
   // All the child directories: the directories referred to by the root and,
-  // recursively, all its children. In order to reconstruct the directory tree,
-  // the client must take the digests of each of the child directories and then
-  // build up a tree starting from the `root`.
-  repeated Directory children = 2;
+  // recursively, all its children. The keys of the map are the hash of the digest
+  // of each directory.
+  map<string, Directory> directories = 4;
```

Comment:
Have we thought about what this means for calculating digests? The ordering of a map in the wire format is undefined so I think the standard serialise + hash approach will be nondeterministic here.

Reply:
I'm pretty sure that doesn't matter - there is nothing using the digest of a Tree as a key explicitly (a Tree is identified solely by the root Directory digest), so the byte representation of the Tree blob should not affect anything from anyone's perspective except internal mechanisms for generating/storing them. Users of the Tree message won't be able to make any assumption as far as byte representations, comparisons between message types should prove equivalent, and blob comparisons are not useful or defined for non-CAS content. Larger transports of directory collections already happen via streaming.

Reply:
The ActionResult does use the digest of the whole Tree message:
```
  // corresponding directory existed after the action completed, a single entry
  // will be present in the output list, which will contain the digest of a
  // [Tree][build.bazel.remote.execution.v2.Tree] message containing the
  // directory tree, and the path equal exactly to the corresponding Action
  // output_directories member.
```
https://github.com/bazelbuild/remote-apis/blob/006a399149dd2dad4eedff247578e94bd957da9c/build/bazel/remote/execution/v2/remote_execution.proto#L922

Reply:
Truth, well played.

The provenance of a Tree in that case might be worth discussing, given that there are only two entities that will likely construct them, and they will not be subject to re-serialization after creation: every one of their uses will be complete.

And one added bonus (for me at least) is that the Java serialization of protobuf maps, while undefined by language, respects a `deterministicSerialization` property of the target output stream. If two maps `.equals()`, then they serialize to the same bytes with this flag enabled. I'll be throwing it on in buildfarm, and recommending so for bazel as well.

Reply:
Agreed. I wonder if there is a possibility of referring to the digest of the directory as you suggested, and this message retaining the map as effectively an optimisation to save future GetTree calls. I'd consider it nicer to have more symmetry between the output and input facets.

Interesting on the Java serialisation; AFAICT there is no similar option for Go (which is most relevant to me, unsure on other languages right now).

Reply:
Looking at maps in https://developers.google.com/protocol-buffers/docs/proto3#maps it's not ideal.  Quoting:

> * Wire format ordering and map iteration ordering of map values is undefined, so you cannot rely on your map items being in a particular order.
> * When generating text format for a .proto, maps are sorted by key. Numeric keys are sorted numerically.
> * When parsing from the wire or when merging, if there are duplicate map keys the last key seen is used. When parsing a map from text format, parsing may fail if there are duplicate keys.
> * If you provide a key but no value for a map field, the behavior when the field is serialized is language-dependent. In C++, Java, and Python the default value for the type is serialized, while in other languages nothing is serialized.

Then there is the backward compatibility section (https://developers.google.com/protocol-buffers/docs/proto3#backwards_compatibility), which leads me to believe that @juergbi 's suggestion may be the viable path forward.  That is:

```
message TreeEntry {
  Digest digest = 1;
  Directory directory = 2;
}

repeated TreeEntry directories = 4;
````

I would suggest to preceed this with the fact that the tree entries should be ordered by digest.  And when parsing that in case of duplicate values, the last encountered wins (rationale: stay aligned with map)

Reply:
The only way to have duplicate entries is to either a) be non-CAS or have different digest functions, or b) literally have replicated directory copies with the same byte representation, so it should be immaterial whether one 'wins.'

Reply:
You're right, stating that the directories MUST be sorted by digest should be sufficient.